sfge/src/test/java/com/salesforce/rules/multiplemassschemalookup/MultipleMassSchemaLookupRuleTest.java

```diff
@@ -0,0 +1,297 @@
+                        4,
+                        "MyClass",
+                        RuleConstants.RepetitionType.LOOP,
+                        ASTConstants.NodeType.FOR_LOOP_STATEMENT));
+    }
+
+    @Test
+    public void testForLoopConditionalOnClassInstance() {
+        String sourceCode[] = {
+            "public class MyClass {\n"
+                    + "   void foo() {\n"
+                    + "       Another[] anotherList = new Another[] {new Another(false), new Another(false)};\n"
+                    + "       for (Another an : anotherList) {\n"
+                    + "           an.exec();\n"
+                    + "       }\n"
+                    + "   }\n"
+                    + "}\n",
+            "public class Another {\n"
+                    + "boolean shouldCheck;\n"
+                    + "Another(boolean b) {\n"
+                    + "   shouldCheck = b;\n"
+                    + "}\n"
+                    + "void exec() {\n"
+                    + "   if (shouldCheck) {\n"
+                    + "       Schema.getGlobalDescribe();\n"
+                    + "   }\n"
+                    + "}\n"
+                    + "}\n"
+        };
+
+        assertViolations(
+                RULE,
+                sourceCode,
+                expect(
+                        8,
+                        RuleConstants.METHOD_SCHEMA_GET_GLOBAL_DESCRIBE,
+                        4,
+                        "MyClass",
+                        RuleConstants.RepetitionType.LOOP,
+                        ASTConstants.NodeType.FOR_EACH_STATEMENT));
+    }
+
+    @Test // TODO: Check if this is a false positive. Static block should get invoked only once
+    public void testLoopFromStaticBlock() {
+        String[] sourceCode = {
+            "public class MyClass {\n"
+                    + "   void foo(String[] objectNames) {\n"
+                    + "       for (Integer i = 0; i < objectNames.size; i++) {\n"
+                    + "           AnotherClass.doNothing();\n"
+                    + "       }\n"
+                    + "   }\n"
+                    + "}\n",
+            "public class AnotherClass {\n"
+                    + "   static {\n"
+                    + "       Schema.getGlobalDescribe();\n"
+                    + "   }\n"
+                    + "   static void doNothing() {\n"
+                    + "   // do nothing \n"
+                    + "   }\n"
+                    + "}\n"
+        };
+
+        assertViolations(
+                RULE,
+                sourceCode,
+                expect(
+                        3,
+                        RuleConstants.METHOD_SCHEMA_GET_GLOBAL_DESCRIBE,
+                        3,
+                        "MyClass",
+                        RuleConstants.RepetitionType.LOOP,
+                        ASTConstants.NodeType.FOR_LOOP_STATEMENT));
+    }
+
+    @Test
+    public void testLoopAfterGgdShouldNotGetViolation() {
```

Comment:
This is a good test. Do we also want one for a loop before a GGD as a counterpart to this one? Also possibly a loop, then a separate loop with a GGD, to verify that it doesn't stop looking after the first loop?

Reply:
Just tried it out and the test fails. All great use cases. Will include them.